### PR TITLE
fix(v2): handle union and optional types

### LINKED ISSUE #939: Figure out handling of Union types
Many features of our current implementation rely on the type hints under the hood to infer behaviour.
This break whenever the type hint is of form `Union` or `Optional`, since those cannot be used in instance and subclass checks.

This is especially problematic, since our type `Tensor` is a Union under the hood, and it is the type used by our pre buit documents. For example, `find()` and `stack()` don't currently work with `Image`.

The simplest workaround is probably to check if a type is union or optional, extract the types inside them, and perform subclass/instance checks on those

### PR DESCRIPTION
**Pull request by JohannesMessner**
Signed-off-by: Johannes Messner <messnerjo@gmail.com>

Goals:

Properly handle `Union` types (such as `Tensor = Union[TorchTensor, NdArray]`) and `Optional` types in stack, find, and bulk access.
Before, all of these cases would throw an exception.

Closes #939

### CODE REVIEW

File: docarray/array/array.py [hunk 1]
```diff
@@ -13,6 +13,8 @@
     Union,
 )
 
+from typing_inspect import is_union_type  # type: ignore
```
Comment (samsja) [inline docarray/array/array.py:16]: you don't need the type ignore if you add to pyprojet toml

File: docarray/util/find.py [hunk 2]
```diff
@@ -1,4 +1,6 @@
-from typing import List, NamedTuple, Optional, Type, Union
+from typing import List, NamedTuple, Optional, Type, Union, get_args
+
+from typing_inspect import is_optional_type, is_union_type  # type: ignore
```
Comment (samsja) [inline docarray/util/find.py:3]: same not needed

File: docarray/array/mixins/attribute.py [hunk 3]
```diff
@@ -21,6 +23,11 @@ def _get_array_attribute(
         in the array like container
         """
         field_type = self.__class__.document_type._get_nested_document_class(field)
+        if is_optional_type(field_type):
+            field_type = get_args(field_type)[0]  # get type inside Optional
+        elif is_union_type(field_type):
+            # determine type based on the first element
+            field_type = type(getattr(self[0], field))
```
Comment (samsja) [inline docarray/array/mixins/attribute.py:30]: so this looks at the data to determine the type right ? Smart idea

Comment (JohannesMessner) [reply docarray/array/mixins/attribute.py:30]: exactly, I don't like it so much actually, but I think it's the only way

**Comment by JohannesMessner** [comment]
Converting to draft bc there are two scenarios still untackled:
- Combination of Optional and Union, which is the case for pre-defined Documents
- stacking Union types should be supported, since pre-defined docs use union type

File: tests/units/array/test_mixins/test_attribute.py [hunk 4]
```diff
@@ -0,0 +1,168 @@
+from typing import Optional, Union
```
Comment (samsja) [inline tests/units/array/test_mixins/test_attribute.py:1]: we don't have mixin in array anymore. we should move this file to somewhere else

Comment (samsja) [reply tests/units/array/test_mixins/test_attribute.py:1]: did u write all of these tests? I don't understand how they are all relevant.

Comment (JohannesMessner) [reply tests/units/array/test_mixins/test_attribute.py]: moved to `test_array`

**Review by samsja — APPROVED** [review]

**Pull request merged**